### PR TITLE
Add thread set_name API.

### PR DESCRIPTION
**Pull request by toots**
This PR adds an API to set a thread's name.

This can be very useful when debugging threads using `gdb`,`lldb` and more.

Unfortunately, the APIs to set thread name are not standard and platform-specific.

The code in this PR is inspired by the [FFmpeg implementation](https://github.com/FFmpeg/FFmpeg/blob/master/libavutil/thread.h#L216) with the windows case added.

At the moment, the implementation does not look for potential errors when setting the name. This is because the error codes might also be inconsistent between platforms and, being a debugging tool, the need to raise a runtime error seems less important.

Please let me know what you think. I can think of the following things:
- Raise error when setting the name fails?
- Add tests? Might need:
- Add get thread name API.

Thanks!

### CODE REVIEW

**Comment by dra27** [comment]
I think the ability to do this is a good idea. Cross-referencing #11264 and #11174 from the multicore merge.

I'd either call the function `set_self_name` (or something making it clear that it refers to the current thread) or, rather more my preference, would be to have the more general `val set_name : t -> string -> unit` which looks like it should be possible on all platforms?

**Comment by toots** [comment]
> I think the ability to do this is a good idea. Cross-referencing #11264 and #11174 from the multicore merge.
> 
> I'd either call the function `set_self_name` (or something making it clear that it refers to the current thread) or, rather more my preference, would be to have the more general `val set_name : t -> string -> unit` which looks like it should be possible on all platforms?

Unfortunately, the macos API doe not allow the choice of `thread_id`, which is why the current function is designed to work inside the thread only.

I'll do the name change!

File: configure.ac
Comment (MisterDA) [inline configure.ac]: [`AC_CHECK_FUNCS([function,...])`](https://www.gnu.org/software/autoconf/manual/autoconf-2.71/html_node/Generic-Functions.html#index-AC_005fCHECK_005fFUNCS-1) defines for each _function_ enumerated in the blank-or-newline-separated argument list `HAVE_function` (in all capitals) if it is available.
I don't think you need special code to test for `prctl`, so this code could be simplified to the following, if you also switch to the `HAVE_` macros.
```suggestion
## Set thread name
AC_CHECK_FUNCS([prctl pthread_setname_np pthread_set_name_np])
```

File: otherlibs/systhreads/st_stubs.c
Comment (MisterDA) [inline otherlibs/systhreads/st_stubs.c:32]: The manual reads:
```
#include <linux/prctl.h>  /* Definition of PR_* constants */
```
is this header not needed?

File: otherlibs/systhreads/st_stubs.c
Comment (MisterDA) [inline otherlibs/systhreads/st_stubs.c]:
```suggestion
  wchar_t *thread_name = caml_stat_strdup_to_utf16(String_val(name));
```

File: otherlibs/systhreads/st_stubs.c
Comment (MisterDA) [inline otherlibs/systhreads/st_stubs.c]: Since 5.3/trunk is currently using winpthreads on both MinGW-w64 and MSVC ports, maybe this should go through winpthreads rather than the WinAPI directly. winpthreads exposes

```c
WINPTHREAD_API pthread_t pthread_self(void);
WINPTHREAD_API int       pthread_setname_np(pthread_t thread, const char *name);
```

File: configure.ac
Comment (toots) [reply configure.ac]: I added the `HAS_` definitions to follow what appears as the naming convention in the OCaml build system.

I put a special case for `prctl` because FFmpeg `configure` script seems to be testing the function presence using the same test code build as proof:

```bash
check_func_headers(){
    log check_func_headers "$@"
    headers=$1
    funcs=$2
    shift 2
    {
        for hdr in $headers; do
            print_include $hdr
        done
        echo "#include <stdint.h>"
        for func in $funcs; do
            echo "long check_$func(void) { return (long) $func; }"
        done
        echo "int main(void) { int ret = 0;"
        # LTO could optimize out the test functions without this
        for func in $funcs; do
            echo " ret |= ((intptr_t)check_$func) & 0xFFFF;"
        done
        echo "return ret; }"
    } | test_ld "cc" "$@" && enable $funcs && enable_sanitized $headers
}

check_func_headers sys/prctl.h prctl
```

I'm suspecting that this is probably because `prctl` might be declared as a `#define` in some systems which would not work with `AC_CHECK_FUNCS` which only check for the function's presence at linkage (see: https://www.gnu.org/software/autoconf/manual/autoconf-2.69/html_node/Generic-Functions.html) which I imagine requires a proper linker symbol.

File: otherlibs/systhreads/st_stubs.c
Comment (toots) [reply otherlibs/systhreads/st_stubs.c:32]: I think the fact that it compiles is a good proof that it isn't 🙂

File: otherlibs/systhreads/st_stubs.c
Comment (toots) [reply otherlibs/systhreads/st_stubs.c]: I'm happy with whichever solution. In general, going for the base OS API seems more robust to future changes.

File: configure.ac
Comment (MisterDA) [reply configure.ac]: > I added the `HAS_` definitions to follow what appears as the naming convention in the OCaml build system.

This is, I believe, purely historical.

> I'm suspecting that this is probably because `prctl` might be declared as a #define in some systems

Thanks for the explanations!
You could use [`AC_CHECK_DECLS`](https://www.gnu.org/software/autoconf/manual/autoconf-2.71/html_node/Generic-Declarations.html#index-AC_005fCHECK_005fDECLS-1)

Comment (toots) [reply configure.ac]: > > I added the `HAS_` definitions to follow what appears as the naming convention in the OCaml build system.
> 
> This is, I believe, purely historical.

Ok. I'm fine going with `HAVE_` if that's the consensus. In my experience, introducing mixed naming convention tends to create confusion.

> > I'm suspecting that this is probably because `prctl` might be declared as a #define in some systems
> 
> Thanks for the explanations! You could use [`AC_CHECK_DECLS`](https://www.gnu.org/software/autoconf/manual/autoconf-2.71/html_node/Generic-Declarations.html#index-AC_005fCHECK_005fDECLS-1)

Done!

File: otherlibs/systhreads/st_stubs.c
Comment (MisterDA) [reply otherlibs/systhreads/st_stubs.c]: I've done a little research and winpthreads doesn't use `SetThreadDescription` because it's too recent. It keeps a local copy of the thread name and throws a particular exception when a debugger is attached. Doing _both_ would cover it all.
Cf [Set a thread name in C/C++](https://learn.microsoft.com/en-us/visualstudio/debugger/tips-for-debugging-threads?view=vs-2022&tabs=csharp#set-a-thread-name-in-cc). I could send a patch to winpthreads…

Comment (toots) [reply otherlibs/systhreads/st_stubs.c]: I'll do both!

File: otherlibs/systhreads/st_stubs.c
Comment (avsm) [inline otherlibs/systhreads/st_stubs.c]: Why not just make the case where it's not implemented a noop instead of raising an exception?

This comes up when compiling to exotic platforms like unikernels, and results in the program raising an exception at start-of-day. But in practise, if there's no interface to set a thread name, then there's no interface to query a thread name either, and you might as well continue the execution of the program. All that happens at this point is that the library runtime has to stub this call out.

Comment (toots) [reply otherlibs/systhreads/st_stubs.c]: Sure thing. Raising an exception is a common pattern when implementations are missing on e.g. unix for windows but since this is for debug and won't impact runtime, it makes sense to silently go ahead. I have added a `stderr` log.

Comment (gasche) [reply otherlibs/systhreads/st_stubs.c]: Emitting `stderr` warnings for something that neither program authors nor program users have much control over is not a terribly popular UI pattern. Within the OCaml runtime, it is controlled by a `caml_runtime_warnings_active()` on the C side, and `Sys.runtime_warnings_enabled ()` on the OCaml side. You should check `caml_runtime_warnings_active()` and not print anything if it is false.

Comment (toots) [reply otherlibs/systhreads/st_stubs.c]: Ok. Just pushed the changes

File: otherlibs/systhreads/st_stubs.c
Comment (gasche) [inline otherlibs/systhreads/st_stubs.c]: \n?

Comment (toots) [inline otherlibs/systhreads/st_stubs.c]:
```suggestion
    fprintf(stderr, "set thread name not implemented\n");
```

Comment (toots) [reply otherlibs/systhreads/st_stubs.c]: yes thanks.

**Comment by gasche** [comment]
This still needs a full review by someone -- maybe @MisterDA or @dra27?

File: otherlibs/systhreads/st_stubs.c
Comment (MisterDA) [reply otherlibs/systhreads/st_stubs.c]: I've [sent a patch](https://sourceforge.net/p/mingw-w64/mailman/message/58829419/) to winpthreads to make it use `SetThreadDescription` in `pthread_setname_np`.

**Review by MisterDA — APPROVED** [review]
I approve this PR.
I don't think `prctl` can be a macro (and thus we could use `AC_CHECK_FUNCS`), it is a syscall wrapper after all.

**Review by gasche — APPROVED** [review]
Approved thanks to @MisterDA's full review.

You should list the reviewers in the change entry, and I think this is good to merge.

**Comment by toots** [comment]
> Approved thanks to @MisterDA's full review.
> 
> You should list the reviewers in the change entry, and I think this is good to merge.

Done. Thanks for all the help with this!

**Comment by dra27** [comment]
> Unfortunately, the macos API doe not allow the choice of `thread_id`, which is why the current function is designed to work inside the thread only.

Ah yes, darn! That said, do we have to restrict the API just because of the one platform? Given the printing of a warning if the function isn't available at all, couldn't we instead expose the general API and display a runtime warning on macOS if it's called with something that isn't Thread.self (and obviously document it in the .mli - cf. the various Windows limitations in unix.mli)? (I'm pushing it slightly given that the limitation of the macOS API appears to cause pain, and we're potentially spreading that pain, rather than containing it!)

**Comment by toots** [comment]
> > Unfortunately, the macos API doe not allow the choice of `thread_id`, which is why the current function is designed to work inside the thread only.
> 
> Ah yes, darn! That said, do we have to restrict the API just because of the one platform? Given the printing of a warning if the function isn't available at all, couldn't we instead expose the general API and display a runtime warning on macOS if it's called with something that isn't Thread.self (and obviously document it in the .mli - cf. the various Windows limitations in unix.mli)? (I'm pushing it slightly given that the limitation of the macOS API appears to cause pain, and we're potentially spreading that pain, rather than containing it!)

I hear you. I'm happy with whatever implementation but, personally, I believe that the burden of having to check what system the code is running on and having to make a different call accordingly in every program using the API will create a long legacy of confusion and friction.

Do we have use-case where it would be absolutely important to be able to name a thread from any other thread using their ID?

At the very least I would suggest an optional thread ID argument and document that thread ID argument is not supported on macos and hope that most users simply call the function with no thread ID argument.

**Comment by MisterDA** [comment]
The Rust stdlib only allows setting the thread name at thread creation, and then retrieving the thread name from within the thread. See [Naming threads](https://doc.rust-lang.org/std/thread/index.html#naming-threads). As an alternative design mimicking this, we could add an optional parameter to `Thread.create`. On macOS, I suppose we'd need to wrap the user-supplied thread function with a preamble setting the thread name.

**Comment by dra27** [comment]
Yes, we could do (as an aside, we already have `caml_thread_t`, so for macOS it could just include a pointer to the string for the initial name and `caml_thread_start` would then be able to apply (and free the memory). Rather than an optional parameter, I'd possibly have `Thread.create_with_name` or some such (i.e. avoid changing the signature of `Thread.create`). The key thing seems to be to have a facility which allows both the thread itself and the thread's creator to set the name, which that would allow.

**Comment by toots** [comment]
However innocuous it looks like, adding an optional parameter to a function already in the module will break all code that is relying on a module type imprint. I've seen that happen with `sedlex` and it can get really hard to track down.

Meanwhile, do we have other calls in the API that have optional debugging params? I think it's good practice to separate debugging calls from actual operational calls. This way they can be opt-in/out easily and also abstracted away when reasoning about the code.

I kinda think we're overthinking this. Do we have use-case in mind? I had one and this API suits it perfectly.

If I had to imagine a use-case where the user requires the ability to name a thread that they have not created I guess it would be a scenario where they want to name e.g. threads from a foreign library.

That seems pretty unlikely. In most reasonable cases, the library would be in charge or asked to provide debugging names for their thread.

Lastly, there's nothing preventing a merge now and the addition of an API with a thread ID later if there is any request for it. This way, we avoid over-engineering it if we don't have a use-case for it right this moment.

**Comment by dra27** [comment]
It wasn’t so much the naming of an arbitrary thread, but I could certainly imagine it being the thread which _creates_ another thread being the point where the name is known, and being forced potentially to pass that information to the thread function itself then being unnecessarily awkward. i.e. the pattern:

```ocaml
let name = Printf.sprintf (* … *) in
let f arg = (* … *) in
let arg = (* … *) in
let th = Thread.create_with_name name f arg
```

vs

```ocaml
let name = Printf.sprintf (* … *) in
let f (name, arg) =
  Thread.set_current_thread_name name;
  (* … *) in
let arg = (* … *) in
let th = Thread.create f (name, arg)
```

But I completely agree that that doesn’t have to be done at the same time as `Thread.set_current_thread_name`

**Comment by gasche** [comment]
My impression is that we could go ahead and merge(squash) the PR in its current state, but I would like to double-check with @dra27 that he agrees. In fact, this is arguably a stdlib PR (more or less), so it may require a second maintainer approval (in addition to mine).

File: otherlibs/systhreads/thread.mli
Comment (dra27) [inline otherlibs/systhreads/thread.mli]: This should mention the fact that a warning can be displayed if runtime warnings are active?

File: otherlibs/systhreads/thread.ml
Comment (dra27) [inline otherlibs/systhreads/thread.ml:96]: Very nittily, but if the function is `set_current_thread_name` can the primitive be `caml_set_current_thread_name`, please?

**Review by dra27 — APPROVED** [review]
I agree with @gasche that this should go in as-is, modulo two nits and it should first run through precheck (which I've tried to schedule as job #989, but apparently not successfully)

File: otherlibs/systhreads/thread.ml
Comment (MisterDA) [reply otherlibs/systhreads/thread.ml:96]: While we're at I find the `current_thread` part a tad verbose. We already have `Thread.self`, I think I prefer the suggested `set_self_name`. Any opinion?

**Comment by gasche** [comment]
It looks like precheck is currently offline due to CI maintenance (maybe @Octachron can confirm?)

File: otherlibs/systhreads/thread.ml
Comment (toots) [reply otherlibs/systhreads/thread.ml:96]: My non-native speaker brain feels like `set_current_thread_name` is more idiomatic/clear than `set_self_thread_name`.

Self is a notion of personal identity or belonging.

The call is made from the perspective of a programmer so, when reading the code, it's more logical to see "Set the current thread name" than "set self thread name" because, when reading the latter as the programmer, this means "set the name of the tread that belong to me".

**Comment by toots** [comment]
> It wasn’t so much the naming of an arbitrary thread, but I could certainly imagine it being the thread which _creates_ another thread being the point where the name is known, and being forced potentially to pass that information to the thread function itself then being unnecessarily awkward. i.e. the pattern:
> 
> ```ocaml
> let name = Printf.sprintf (* … *) in
> let f arg = (* … *) in
> let arg = (* … *) in
> let th = Thread.create_with_name name f arg
> ```
> 
> vs
> 
> ```ocaml
> let name = Printf.sprintf (* … *) in
> let f (name, arg) =
>   Thread.set_current_thread_name name;
>   (* … *) in
> let arg = (* … *) in
> let th = Thread.create f (name, arg)
> ```
> 
> But I completely agree that that doesn’t have to be done at the same time as `Thread.set_current_thread_name`

Ha that makes sense thanks!

File: otherlibs/systhreads/thread.mli
Comment (toots) [reply otherlibs/systhreads/thread.mli]: Addressed!

File: otherlibs/systhreads/thread.ml
Comment (toots) [reply otherlibs/systhreads/thread.ml:96]: Renamed applied!

**Comment by toots** [comment]
I have addressed the last batch to comments, thanks y'all!

File: otherlibs/systhreads/st_stubs.c
Comment (MisterDA) [reply otherlibs/systhreads/st_stubs.c]: The patch's been merged!
https://github.com/mingw-w64/mingw-w64/commit/04ec40f5bcdd3966c1bfac7e64a8bacfdc9671bd

**Comment by toots** [comment]
Hi there! Any update on this?

**Comment by gasche** [comment]
I just started a precheck build at https://ci.inria.fr/ocaml/job/precheck/990/ . If it passes I'll merge.

**Comment by gasche** [comment]
I get a build failure on the [ocaml-mingw-32](https://ci.inria.fr/ocaml/job/precheck/990/flambda=false,label=ocaml-mingw-32/console) machine of `precheck`:

```
st_stubs.c: In function 'caml_set_current_thread_name':
st_stubs.c:976:3: error: implicit declaration of function 'SetThreadDescription'; did you mean 'SetThreadDesktop'? [-Werror=implicit-function-declaration]
  976 |   SetThreadDescription(GetCurrentThread(), thread_name);
      |   ^~~~~~~~~~~~~~~~~~~~
      |   SetThreadDesktop
```

**Comment by MisterDA** [comment]
This symbol was added in https://github.com/mingw-w64/mingw-w64/commit/3a137bd87ea83d174f4a225568b75cabdbc70852, released in MinGW-w64 12.0 (2024-05-29). The CI machine is using MinGW-w64 9.0 (2021-05-22). As a workaround, we could add the declaration for versions of MinGW-w64 before 12 as such:

```c
#if defined(__MINGW64_VERSION_MAJOR) && __MINGW64_VERSION_MAJOR < 12
WINBASEAPI HRESULT WINAPI SetThreadDescription (HANDLE hThread, PCWSTR lpThreadDescription);
#endif
```

or we could restrict the build to MinGW-w64 12.0 (as a side note, the winpthreads patches required for the MSVC port for the OCaml 5.3 branch were released as part of MinGW-w64 12.0).

**Comment by gasche** [comment]
No opinion here. Maybe @toots has an opinion on what is best (but I suppose tweaking the CI is something someone else would do), otherwise of course @dra27 could provide guidance.

File: configure.ac
Comment (MisterDA) [inline configure.ac]: This only detects whether the _linker_ finds the symbol. As MinGW-w64 9.0 doesn't provide the *declaration*, I suspect that the configure test will succeed, but compilation will fail, as the symbol won't be found when building this file. I suggest you declare the symbol too for versions prior to MinGW-w64 12.0.

**Review by MisterDA — CHANGES_REQUESTED** [review]

File: configure.ac
Comment (toots) [reply configure.ac]: Good catch thanks. I'm trying to be more factual than relying on versions. I added a test that checks if the symbol is declared in the expected header.

If not but the symbol exists, then we define it. That should stick to the problem at hand and fix all possible situations the way it's supposed to be.

File: configure.ac
Comment (MisterDA) [inline configure.ac]:
```suggestion
  [AC_DEFINE([HAS_DECL_PRCTL] , [1])],
```

File: configure.ac
Comment (MisterDA) [inline configure.ac]:
```suggestion
## SetThreadDescription
AC_CHECK_FUNCS(
  [SetThreadDescription],
  [AC_DEFINE([HAS_SETTHREADDESCRIPTION], [1])
  AC_CHECK_DECL(
    [SetThreadDescription],
    [AC_DEFINE([HAS_DECL_SETTHREADDESCRIPTION], [1])],
    [#include <processthreadsapi.h>])])
```

File: otherlibs/systhreads/st_stubs.c
Comment (MisterDA) [inline otherlibs/systhreads/st_stubs.c]:
```suggestion
#  if defined(HAS_SETTHREADDESCRIPTION) && !defined(HAS_DECL_SETTHREADDESCRIPTION)
```

File: otherlibs/systhreads/st_stubs.c
Comment (MisterDA) [inline otherlibs/systhreads/st_stubs.c]:
```suggestion
#elif defined(HAS_DECL_PRCTL)
```

File: otherlibs/systhreads/st_stubs.c
Comment (MisterDA) [inline otherlibs/systhreads/st_stubs.c]:
```suggestion
#  if defined(HAS_SETTHREADDESCRIPTION)
```

File: otherlibs/systhreads/st_stubs.c
Comment (MisterDA) [inline otherlibs/systhreads/st_stubs.c]:
```suggestion
#elif defined(HAS_DECL_PRCTL)
```

File: runtime/caml/s.h.in
Comment (MisterDA) [inline runtime/caml/s.h.in]:
```suggestion
#undef HAS_DECL_PRCTL
```

File: runtime/caml/s.h.in
Comment (MisterDA) [inline runtime/caml/s.h.in]:
```suggestion
#undef HAS_SETTHREADDESCRIPTION

#undef HAS_DECL_SETTHREADDESCRIPTION
```

**Review by MisterDA — APPROVED** [review]
LGTM, now we're covered. I'm just suggesting another way of naming the macros that's more in line with what default Autoconf would use.

**Comment by toots** [comment]
I agree on keeping conventions. Renames applied!

**Review by MisterDA — APPROVED** [review]

File: configure.ac
Comment (MisterDA) [inline configure.ac]:
```suggestion
  [AC_DEFINE([HAS_DECL_SETTHREADDESCRIPTION], [1])], []
  [#include <processthreadsapi.h>])
```

The include was mistakenly placed in the _action-if-not-found_ block. See [`AC_CHECK_DECL`](https://www.gnu.org/software/autoconf//manual/autoconf-2.71/html_node/Generic-Declarations.html#index-AC_005fCHECK_005fDECL-1).

File: otherlibs/systhreads/st_stubs.c
Comment (MisterDA) [inline otherlibs/systhreads/st_stubs.c]:
```suggestion
#  if defined(HAS_SETTHREADDESCRIPTION) && \
      !defined(HAS_DECL_SETTHREADDESCRIPTION)
```
I think this is fine without the extra line…

**Comment by gasche** [comment]
I ran a new [precheck build](https://ci.inria.fr/ocaml/job/precheck/991/) and unfortunately this still fails with the msvc-64 CI machine:

```
  CC st_stubs.b.obj
st_stubs.c
st_stubs.c(28): error C2220: l'avertissement suivant est traité comme une erreur
st_stubs.c(28): warning C4273: 'SetThreadDescription' : liaison DLL incohérente
C:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\um\processthreadsapi.h(1422): note: voir la définition précédente de 'SetThreadDescription'
make[4]: *** [Makefile:84: st_stubs.b.obj] Error 2
```

The configure logs have the following:

```
checking for SetThreadDescription... yes
checking whether SetThreadDescription is declared... no
```

File: configure.ac
Comment (MisterDA) [inline configure.ac]:
```suggestion
  [[#define WIN32_LEAN_AND_MEAN
    #include <windows.h>
    #include <processthreadsapi.h>]])
```

The last error! unfortunately the `processthreadspi.h` header isn't self-contained.

**Comment by gasche** [comment]
The third iteration of precheck build ([#992](https://ci.inria.fr/ocaml/job/precheck/992/)) is now successful, thanks @toots and @MisterDA! Merging now.

**Comment by toots** [comment]
Awesome y'all thanks for all y'all patience and help with this!

**Comment by MisterDA** [comment]
The commits could have been squashed… the history is kind of messy now. For next time?

**Comment by gasche** [comment]
Yes, apologies for clicking the wrong button.

**Comment by toots** [comment]
This could be force through the GH settings.

**Pull request merged**
